internal/security/bootstrapper/command/setupacl/command.go

```diff
@@ -0,0 +1,364 @@
+)
+
+const (
+	// the command name for setting up registry's ACL
+	CommandName string = "setupRegistryACL"
+
+	consulGetLeaderAPI         = "/v1/status/leader"
+	consulACLBootstrapAPI      = "/v1/acl/bootstrap"
+	consulConfigAccessVaultAPI = "/v1/consul/config/access"
+	consulLegacyACLModeError   = "The ACL system is currently in legacy mode"
+	defaultRetryTimeout        = 30 * time.Second
+	emptyLeader                = `""`
+	emptyToken                 = ""
+)
+
+type cmd struct {
+	waitGroup     *sync.WaitGroup
+	loggingClient logger.LoggingClient
+	client        internal.HttpCaller
+	configuration *config.ConfigurationStruct
+
+	// internal state
+	errs         chan error
+	retryTimeout time.Duration
+}
+
+// NewCommand creates a new cmd and parses through options if any
+func NewCommand(
+	_ context.Context,
+	_ *sync.WaitGroup,
+	lc logger.LoggingClient,
+	conf *config.ConfigurationStruct,
+	args []string) (interfaces.Command, error) {
+	cmd := cmd{
+		waitGroup:     &sync.WaitGroup{},
+		loggingClient: lc,
+		client:        pkg.NewRequester(lc).Insecure(),
+		configuration: conf,
+		retryTimeout:  defaultRetryTimeout,
+	}
+	var dummy string
+
+	flagSet := flag.NewFlagSet(CommandName, flag.ContinueOnError)
+	flagSet.StringVar(&dummy, "confdir", "", "") // handled by bootstrap; duplicated here to prevent arg parsing errors
+
+	err := flagSet.Parse(args)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse command: %s: %w", strings.Join(args, " "), err)
+	}
+
+	return &cmd, nil
+}
+
+// Execute implements Command and runs this command
+// command setupRegistryACL sets up the ACL system of the registry, Consul in this case, preparing for generating
+// Consul's agent tokens later on
+func (c *cmd) Execute() (statusCode int, err error) {
+	c.loggingClient.Infof("Security bootstrapper running %s", CommandName)
+
+	if err := c.waitForNonEmptyConsulLeader(); err != nil {
+		return interfaces.StatusCodeExitWithError, fmt.Errorf("failed to wait for Consul leader: %v", err)
+	}
+
+	var bootstrapTokenInfo *BootStrapACLTokenInfo
+	bootstrapTokenInfo, err = c.callConsulACLBootstrapAPI()
+	if err != nil {
+		// although we have a leader, but it is a very very rare chance that we could hit an error on legacy mode
+		// here we will sleep a bit of time and then retry once if there is error on Legacy ACL type of message
+		// because Consul is still on its way to initialize the new ACL system internally
+		// for the details of this issue, see related issue on Consul's Github website:
+		// https://github.com/hashicorp/consul/issues/5218#issuecomment-457212336
+		if !strings.Contains(err.Error(), consulLegacyACLModeError) {
+			// other type of ACL bootstrapping error, cannot continue
+			return interfaces.StatusCodeExitWithError, fmt.Errorf("failed to bootstrap registry's ACL: %v", err)
+		}
+
+		c.loggingClient.Warnf("found Consul still in ACL legacy mode, will retry once again: %v", err)
+		time.Sleep(5 * time.Second)
```

Comment:
5 seconds seems like a very long time to wait. I would prefer 1 second with 5 retries.

Reply:
not really, acl bootstrapping is kind of expensive process, so less tries is better actually

Reply:
@bnevis-i , thoughts on this long timeout?

Reply:
Looks OK to me.  I set up a HA Vault before and it takes quite a number of seconds for leader election, etc.